pkg/machinehealthcheck/machinehealthcheck.go

```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 
 		By("Creating a MachineHealthCheck resource")
 		mhcParams := framework.MachineHealthCheckParams{
-			Name:   machineSetParams.Name,
-			Labels: machineSetParams.Labels,
+			Name:   machineSet.Name,
+			Labels: machineSet.Labels,
 			Conditions: []mapiv1beta1.UnhealthyCondition{
 				{
 					Type:    E2EConditionType,
 					Status:  corev1.ConditionTrue,
 					Timeout: "1s",
 				},
 			},
+			MaxUnhealthy: &maxUnhealthy,
 		}
 
-		mhc, err := framework.CreateMHC(client, mhcParams)
+		machinehealthcheck, err = framework.CreateMHC(client, mhcParams)
 		Expect(err).ToNot(HaveOccurred())
-		Expect(mhc).ToNot(BeNil())
+		Expect(machinehealthcheck).ToNot(BeNil())
 
-		By("Verifying the matching Machine is deleted")
-		framework.WaitForMachineDelete(client, machine)
+		By("Waiting for each unhealthy machine to be deleted")
+		framework.WaitForMachinesDeleted(client, unhealthyMachines...)
+
+		By("Waiting for MachineDeleted event from MachineHealthCheck for each unhealthy machine")
+		for _, machine := range unhealthyMachines {
+			err = framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")
+			Expect(err).ToNot(HaveOccurred())
+		}
 
 		By("Verifying the MachineSet recovers")
 		framework.WaitForMachineSet(client, machineSet.GetName())
+	})
 
-		By("Deleting the MachineHealthCheck resource")
-		err = client.Delete(context.Background(), mhc)
+	It("should not remediate larger number of unhealthy machines then maxUnhealthy", func() {
+		selector := machineSet.Spec.Selector
+		machines, err := framework.GetMachines(client, &selector)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machines).ToNot(BeEmpty())
+
+		By("Setting unhealthy conditions on machine nodes, but exceding maxUnhealthy threshold")
+		unhealthyMachines := machines[:maxUnhealthy+1]
+		for _, machine := range unhealthyMachines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			err = framework.AddNodeCondition(client, node, nodeCondition)
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		By("Creating a MachineHealthCheck resource")
+		mhcParams := framework.MachineHealthCheckParams{
+			Name:   machineSet.Name,
+			Labels: machineSet.Labels,
+			Conditions: []mapiv1beta1.UnhealthyCondition{
+				{
+					Type:    E2EConditionType,
+					Status:  corev1.ConditionTrue,
+					Timeout: "1s",
+				},
+			},
+			MaxUnhealthy: &maxUnhealthy,
+		}
+
+		machinehealthcheck, err = framework.CreateMHC(client, mhcParams)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machinehealthcheck).ToNot(BeNil())
+
+		By("Waiting for RemediationRestricted event from MachineHealthCheck")
+		err = framework.WaitForEvent(client, "MachineHealthCheck", machineSet.Name, "RemediationRestricted")
 		Expect(err).ToNot(HaveOccurred())
-	})
 
-	// It("should respect maxUnhealthy", func() {
-	// 	// TODO(bison): This.
-	// })
+		By("Ensuring none of the unhealthy machines are deleted")
+		untouchedMachines, err := framework.GetMachines(client, &selector)
```

Comment:
Perhaps healthy rather than untouched, untouched seems a bit of a weird description to me 😅